### PR TITLE
WIP: Port SpatialAveragePooling CPU to ATen

### PR DESCRIPTION
**Pull request by hameerabbasi**
Not ready for review.

### CODE REVIEW

**Comment by hameerabbasi** [comment]
https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/SpatialAveragePooling.c#L8-L55

This functions checks the shape, and calculates the output's shape as well. I'm not sure if we should refactor it to return the output shape as well.

I'm not sure where the output is being allocated in the first place, since it should be the same size.

File: aten/src/ATen/native/SpatialAveragePooling.cpp [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+          outputWidth,
+          nbatch,
+          count_include_pad,
+          kernel_size,
+          stride_size,
+          pad_size);
+    });
+}
+
+} // namespace
```
Comment (skrah) [inline aten/src/ATen/native/SpatialAveragePooling.cpp]: I assume you were going to add the wrapper functions here. Can you add the functions that wrap e.g. `spatial_avg_pool2d_out_cpu_template` see for examples `AdaptiveMaxPool2d.cpp` or `FractionalMaxPool2d.cpp`?

Then it's easier to talk about the `output` shape inference.

**Comment by hameerabbasi** [comment]
I'm getting the following errors when linking:

```
[build] Undefined symbols for architecture x86_64:
[build]   "at::native::avg_pool2d_cpu(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool)", referenced from:
[build]       at::CPUByteType::avg_pool2d(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUByteType.cpp.o
[build]       at::CPUCharType::avg_pool2d(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUCharType.cpp.o
[build]       at::CPUDoubleType::avg_pool2d(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUDoubleType.cpp.o
[build]       at::CPUFloatType::avg_pool2d(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUFloatType.cpp.o
[build]       at::CPUIntType::avg_pool2d(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUIntType.cpp.o
[build]       at::CPULongType::avg_pool2d(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPULongType.cpp.o
[build]       at::CPUQInt8Type::avg_pool2d(at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUQInt8Type.cpp.o
[build]       ...
[build]   "at::native::avg_pool2d_out_cpu(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool)", referenced from:
[build]       at::CPUByteType::avg_pool2d_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUByteType.cpp.o
[build]       at::CPUCharType::avg_pool2d_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUCharType.cpp.o
[build]       at::CPUDoubleType::avg_pool2d_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUDoubleType.cpp.o
[build]       at::CPUFloatType::avg_pool2d_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUFloatType.cpp.o
[build]       at::CPUIntType::avg_pool2d_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUIntType.cpp.o
[build]       at::CPULongType::avg_pool2d_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPULongType.cpp.o
[build]       at::CPUQInt8Type::avg_pool2d_out(at::Tensor&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUQInt8Type.cpp.o
[build]       ...
[build]   "at::native::avg_pool2d_backward_cpu(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool)", referenced from:
[build]       at::CPUByteType::avg_pool2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUByteType.cpp.o
[build]       at::CPUCharType::avg_pool2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUCharType.cpp.o
[build]       at::CPUDoubleType::avg_pool2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUDoubleType.cpp.o
[build]       at::CPUFloatType::avg_pool2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUFloatType.cpp.o
[build]       at::CPUIntType::avg_pool2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUIntType.cpp.o
[build]       at::CPULongType::avg_pool2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPULongType.cpp.o
[build]       at::CPUQInt8Type::avg_pool2d_backward(at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUQInt8Type.cpp.o
[build]       ...
[build]   "at::native::avg_pool2d_backward_out_cpu(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool)", referenced from:
[build]       at::CPUByteType::avg_pool2d_backward_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUByteType.cpp.o
[build]       at::CPUCharType::avg_pool2d_backward_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUCharType.cpp.o
[build]       at::CPUDoubleType::avg_pool2d_backward_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUDoubleType.cpp.o
[build]       at::CPUFloatType::avg_pool2d_backward_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUFloatType.cpp.o
[build]       at::CPUIntType::avg_pool2d_backward_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUIntType.cpp.o
[build]       at::CPULongType::avg_pool2d_backward_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPULongType.cpp.o
[build]       at::CPUQInt8Type::avg_pool2d_backward_out(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::ArrayRef<long long>, c10::ArrayRef<long long>, c10::ArrayRef<long long>, bool, bool) const in CPUQInt8Type.cpp.o
[build]       ...
[build] ld: symbol(s) not found for architecture x86_64
[build] clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

File: aten/src/ATen/native/SpatialAveragePooling.cpp (hunk 1 above)
Comment (hameerabbasi) [reply aten/src/ATen/native/SpatialAveragePooling.cpp]: Thanks!

**Comment by hameerabbasi** [comment]
@skrah There seems to be a small mistake I'm making, but this still won't build. Is it possible for you to take a look?

**Comment by skrah** [comment]
@hameerabbasi I can take a look if you rebase on master. There's a conflict in one of the crucial sections.

**Comment by hameerabbasi** [comment]
It seems this was solved by https://github.com/pytorch/pytorch/commit/4864000e55150955355778744ba448a03a57f7a1

**Comment by skrah** [comment]
I still get a conflict in `native_functions.yaml`.  You can ask `pytorchbot` to rebase, but I think it will fail.

**Comment by hameerabbasi** [comment]
I should have been more verbose — These functions were ported by that commit I mentioned earlier. This PR is irrelevant.

**Comment by rgommers** [comment]
> I should have been more verbose — These functions were ported by that commit I mentioned earlier. This PR is irrelevant.

@hameerabbasi I'm not sure that's right. That commit only adds an MKL-DNN based version, and raises an error when MKL is not present. PyTorch can be compiled without MKL, and MKL also doesn't support all CPU types I believe. There's a bit of discussion on adding MKL-DNN [here](https://github.com/pytorch/pytorch/issues/16038#issuecomment-466308142), I can't find too much else right now.

**Comment by skrah** [comment]
I agree with @rgommers: `SpatialAveragePooling.c` is still in `THNN` in the latest master.  I think the title of this PR is wrong. :)

**Comment by hameerabbasi** [comment]
@skrah I rebased, as requested, could you take a look?

**Comment by skrah** [comment]
But the branch still has conflicts, which is also what the grey GitHub icon says.

```
$ git pull upstream port-adaptive-pool-2d-cpu
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Total 23 (delta 14), reused 14 (delta 14), pack-reused 9
Unpacking objects: 100% (23/23), done.
From https://github.com/pytorch/pytorch
 * branch            port-adaptive-pool-2d-cpu -> FETCH_HEAD
 + 61e414b...0dd27e4 port-adaptive-pool-2d-cpu -> upstream/port-adaptive-pool-2d-cpu  (forced update)
warning: Failed to merge submodule third_party/fbgemm (commits don't follow merge-base)
Auto-merging third_party/fbgemm
CONFLICT (submodule): Merge conflict in third_party/fbgemm
Auto-merging aten/src/ATen/native/native_functions.yaml
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by skrah** [comment]
Also the diff includes ` third_party/eigen` etc., so that would need to be removed before merging into master anyway.

**Comment by skrah** [comment]
If it is faster, I'd just open a new PR, which sidesteps all the git merge/rebase fun. :)

**Comment by hameerabbasi** [comment]
Weird, the rebase introduced additional files, which has never happened before (submodules, yay?). Anyway, I removed those and pushed. :)

File: aten/src/ATen/native/SpatialAveragePooling.cpp [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+
+Tensor& avg_pool2d_out_cpu(
+  Tensor& input,
+  IntList kernel_size,
+  IntList stride,
+  IntList padding,
+  bool ceil_mode,
+  bool count_include_pad,
+  Tensor& output)
```
Comment (skrah) [inline aten/src/ATen/native/SpatialAveragePooling.cpp]: These signatures need to match the generated ones in `build/aten/src/ATen/NativeFunctions.h`. The `out` argument is moved to the front in that header and does not necessarily match the order in `native_functions.yaml`.

Comment (skrah) [reply aten/src/ATen/native/SpatialAveragePooling.cpp]: I'm pretty sure this is why you get build failures.

**Comment by hameerabbasi** [comment]
I fixed the signatures, but now it fails to link `avg_pool2d_backward_out`, which isn't even in the YAML file nor in my PR.

**Pull request closed without merging**